bitcoin/src/blockdata/witness.rs

```diff
@@ -147,19 +147,26 @@ crate::internal_macros::define_extension_trait! {
         ///
         /// See [`Script::is_p2tr`] to check whether this is actually a Taproot witness.
         fn tapscript(&self) -> Option<&Script> {
-            self.last().and_then(|last| {
-                // From BIP341:
-                // If there are at least two witness elements, and the first byte of
-                // the last element is 0x50, this last element is called annex a
-                // and is removed from the witness stack.
-                if self.len() >= 3 && last.first() == Some(&TAPROOT_ANNEX_PREFIX) {
-                    self.nth(self.len() - 3).map(Script::from_bytes)
-                } else if self.len() >= 2 {
-                    self.nth(self.len() - 2).map(Script::from_bytes)
-                } else {
-                    None
-                }
-            })
+            let has_annex = self.taproot_annex().is_some();
+
+            let minimum_script_path_spend_size = if has_annex {
+              3
+            } else {
+              2
+            };
+
+            if self.len() < minimum_script_path_spend_size {
+              return None;
+            }
+
+            let tapscript_index = if has_annex {
+              self.len() - 3
+            } else {
+              self.len() - 2
+            };
```

Comment:
In 96d5d2771542b5c172985f85ea833d14604a4de0:

You can collapse this if branch to just

```rust
let tapscript_index = self.len() - minimum_script_path_spend_size;
```

and even collapse it into the above check

```rust
let tapscript_index = self.len().checked_sub(minimum_script_path_spend_size)?;
```

Reply:
If you prefer it, I can definitely do that. However, I actually think that being verbose is an advantage here. A subtle bug was introduced into the previous version, and the logic is a little complicated, so a version which is more explicit might be better. I would almost say it should be:

```rust
let last = self.len() - 1;

let tapscript_index = if has_annex {
  last - 2
} else {
  last - 1
};
```

Which is more verbose, but doesn't conflate subtracting one from the length to get a valid index, and subtracting one or two from the length to get the index of the tapscript, skipping both the control block and the annex, or skipping just the control block.

Reply:
Let me know what you think is best!

Reply:
Good observations. Yeah, let's add a `last` variable as you suggest rather than trying to find clever one-liners.

Reply:
Sounds good, made it even more verbose!